<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Introduction &mdash; ST Pipeline 1.3.3 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.3.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="ST Pipeline 1.3.3 documentation" href="index.html" />
    <link rel="next" title="Installing the Spatial Transcriptomics pipeline" href="installation.html" />
    <link rel="prev" title="Welcome to ST Pipeline&#39;s documentation!" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="installation.html" title="Installing the Spatial Transcriptomics pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to ST Pipeline&#39;s documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ST Pipeline 1.3.3 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">Â¶</a></h1>
<p>The ST Pipeline contains the tools and scripts needed to process
and analyze the raw files generated with the Spatial Transcriptomics
method in FASTQ format to generated datasets for down-stream analysis.
The ST pipeline can also be used to process single cell data as
long as a file with barcodes identifying each cell is provided.</p>
<p>The ST Pipeline has been optimized for speed, robustness and it is very
easy to use with many parameters to adjust all the settings.</p>
<p>The following files/parameters are required:</p>
<ul>
<li><p class="first">FASTQ files (Read 1 containing the spatial information and the UMI
and read 2 containing the genomic sequence)</p>
</li>
<li><p class="first">A genome index generated with STAR</p>
</li>
<li><p class="first">An annotation file in GTF or GFF format</p>
</li>
<li><dl class="first docutils">
<dt>The file containing the barcodes and array coordinates</dt>
<dd><p class="first last">(look at the folder &quot;ids&quot; and chose the correct one).
Basically this file contains 3 columns (BARCODE, X and Y),
so if you provide this file with barcodes identinfying cells (for example),
the ST pipeline can be used for single cell data.</p>
</dd>
</dl>
</li>
<li><p class="first">A name for the dataset</p>
</li>
</ul>
<p>The ST pipeline has multiple parameters mostly related to trimming,
mapping and annotation but generally the default values are good enough.
You can see a full description of the parameters
typing &quot;st_pipeline_run.py --help&quot; after you have installed the ST pipeline.</p>
<p>The input FASTQ files can be given in gzip/bzip format as well.</p>
<p>Basically what the ST pipeline does is:</p>
<ul>
<li><dl class="first docutils">
<dt>Quality trimming (read 1 and read 2):</dt>
<dd><ul class="first last simple">
<li>Remove low quality bases</li>
<li>Sanity check (reads same length, reads order, etc..)</li>
<li>Check quality UMI</li>
<li>Remove artifacts (PolyT, PolyA, PolyG, PolyN and PolyC) of user defined length</li>
<li>Check for AT and GC content</li>
<li>Discard reads with a minimum number of bases of that failed any of the checks above</li>
</ul>
</dd>
</dl>
</li>
<li><p class="first">Contamimant filter e.x. rRNA genome (Optional)</p>
</li>
<li><p class="first">Mapping with STAR (only read 2)</p>
</li>
<li><p class="first">Demultiplexing with [Taggd](<a class="reference external" href="https://github.com/SpatialTranscriptomicsResearch/taggd">https://github.com/SpatialTranscriptomicsResearch/taggd</a>) (only read 1)</p>
</li>
<li><p class="first">Keep reads (read 2) that contain a valid barcode and are correctly mapped</p>
</li>
<li><p class="first">Annotate the reads with htseq-count (slightly modified version)</p>
</li>
<li><p class="first">Group annotated reads by barcode(spot position) and gene to get a read count</p>
</li>
<li><p class="first">In the grouping/counting only unique molecules (UMIs) are kept.</p>
</li>
</ul>
<p>You can see a graphical more detailed description of the workflow in the documents workflow.pdf and workflow_extended.pdf</p>
<p>The output will be a matrix of counts (genes as columns, spots as rows),
a BED file containing the transcripts (Read name, coordinate, gene, etc..), and a JSON
file with useful stats.
The ST pipeline will also output a log file with useful information.</p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to ST Pipeline's documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="installation.html"
                        title="next chapter">Installing the Spatial Transcriptomics pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/intro.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="installation.html" title="Installing the Spatial Transcriptomics pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to ST Pipeline&#39;s documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ST Pipeline 1.3.3 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Jose Fernandez Navarro.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div>
  </body>
</html>